testbed/correctness/correctness_test.go

```diff
@@ -0,0 +1,231 @@
+	case "opencensus":
+		sender = testbed.NewOCTraceDataSender(testbed.DefaultHost, testbed.GetAvailablePort(t))
+	case "jaeger":
+		sender = testbed.NewJaegerGRPCDataSender(testbed.DefaultHost, testbed.GetAvailablePort(t))
+	case "zipkin":
+		sender = testbed.NewZipkinDataSender(testbed.DefaultHost, testbed.GetAvailablePort(t))
+	default:
+		t.Errorf("unknown receiver type: %s", receiver)
+	}
+	return sender
+}
+
+func constructReceiver(t *testing.T, exporter string) testbed.DataReceiver {
+	var receiver testbed.DataReceiver
+	switch exporter {
+	case "otlp":
+		receiver = testbed.NewOTLPDataReceiver(testbed.GetAvailablePort(t))
+	case "opencensus":
+		receiver = testbed.NewOCDataReceiver(testbed.GetAvailablePort(t))
+	case "jaeger":
+		receiver = testbed.NewJaegerDataReceiver(testbed.GetAvailablePort(t))
+	case "zipkin":
+		receiver = testbed.NewZipkinDataReceiver(testbed.GetAvailablePort(t))
+	default:
+		t.Errorf("unknown exporter type: %s", exporter)
+	}
+	return receiver
+}
+
+func loadPictOutputPipelineDefs(fileName string) ([]PipelineDef, error) {
+	file, err := os.Open(filepath.Clean(fileName))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	defs := make([]PipelineDef, 0)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		s := strings.Split(scanner.Text(), "\t")
+		if "Receiver" == s[0] {
+			continue
+		}
+
+		var aDef PipelineDef
+		aDef.receiver, aDef.exporter = s[0], s[1]
+		defs = append(defs, aDef)
+	}
+
+	return defs, err
+}
+
+func testWithGoldenDataset(
+	t *testing.T,
+	sender testbed.DataSender,
+	receiver testbed.DataReceiver,
+	resourceSpec testbed.ResourceSpec,
+	processors map[string]string,
+) {
+	resultDir, err := filepath.Abs(path.Join("results", t.Name()))
+	require.NoError(t, err)
+
+	dataProvider := testbed.NewGoldenDataProvider(
+		"../../internal/goldendataset/testdata/generated_pict_pairs_traces.txt",
+		"../../internal/goldendataset/testdata/generated_pict_pairs_spans.txt",
+		161803)
+	factories, err := defaultcomponents.Components()
+	assert.NoError(t, err)
+	runner := testbed.NewInProcessPipeline(factories)
+	config := createConfigFile(t, sender, receiver, resultDir, processors)
+	_, cfgErr := runner.PrepareConfig(config)
+	assert.NoError(t, cfgErr)
+	tc := testbed.NewTestCase(
+		t,
+		dataProvider,
+		sender,
+		receiver,
+		runner,
+		correctnessResults,
+	)
+	defer tc.Stop()
+
+	tc.SetResourceLimits(resourceSpec)
+	tc.StartBackend()
+	tc.StartAgent()
+
+	tc.StartLoad(testbed.LoadOptions{
+		DataItemsPerSecond: 10000,
+		ItemsPerBatch:      100,
+	})
+
+	tc.Sleep(tc.Duration)
+
+	tc.StopLoad()
+
+	tc.WaitFor(func() bool { return tc.LoadGenerator.DataItemsSent() == tc.MockBackend.DataItemsReceived() },
+		"all data items received")
+
+	tc.StopAgent()
+
+	tc.ValidateData()
+}
+
+func createConfigFile(t *testing.T, sender testbed.DataSender, receiver testbed.DataReceiver, resultDir string,
```

Comment:
Is this a copy of createConfigFile from scenarios.go or is different from it? If it is a copy perhaps extract as testbed helper func.

Reply:
This is a different configuration than used by perf tests. Generating string and converting to Config because functionality does not currently exist to write out Config back to a string.